### DISCUSSION TITLE
A2DP -> Mono -> I2S distorted

### DISCUSSION THREAD

**Discussion by fscii** [pschatzmann/arduino-audio-tools#1602] [Q&A] [answered]
I believe had a mono example working a while back, went on vaction, did various number of git-pulls and now there's horrible sound (distorted and about an octave or more lower) when using mono.  Works fine in stereo.  Any advice?

PS: I saw this said only change channels = 1;

(https://github.com/pschatzmann/arduino-audio-tools/discussions/836)

`#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

I2SStream i2s;
BluetoothA2DPSink a2dp_sink(i2s);                  // I2S Source -> A2DP Sink

void setup() {

    // I2S DAC settings
    auto cfg         = i2s.defaultConfig(TX_MODE); 
   // cfg.i2s_format   = I2S_STD_FORMAT;            // PCM5102A seems obvlivious to format

    // ESP32 pins                                  // library default bck=14, ws=15, data=22

    cfg.pin_bck     = 15;  //12 - 15
    cfg.pin_ws      = 27;  //14 - 27
    cfg.pin_data   = 26;  //27 - 26

    cfg.channels = 1;  // Mono

    i2s.begin(cfg);                                // Begin source before sink
    a2dp_sink.start("I2S_Mono_Test");
}

void loop() {
}`

I tried both yellow and purple PCM5102 DAC boards on this page w/same restults:
https://github.com/pschatzmann/arduino-audio-tools/wiki/External-DAC

Code works fine in stereo (cfg.channels = 2;) but not mono.

**Comment by pschatzmann** [comment, marked as answer]
In Arduino ESP32 3.0 core the I2S has changed and might therefore behave slightly differently.
A2DP provides stereo data, so only specifying channels = 1 in I2S will not work.

You can use a2dp_sink.[set_mono_downmix](https://pschatzmann.github.io/ESP32-A2DP/html/class_bluetooth_a2_d_p_sink.html#a624040cce89a4a2f66495f57db6c1457) (bool enabled) to output the same (mono) signal on the 2 i2s channels.

If you want to limit the I2S signal to 1 output pin you need to [convert the number of channels first](https://github.com/pschatzmann/arduino-audio-tools/wiki/Converting-the-Data-Format)!

**Reply by fscii** [reply]
Thank you for clearing that up.  downmix works good!  I checked out the example you showed (signal generator) and also this one:
[https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-formatconverter-i2s/streams-generator-formatconverter-i2s.ino](url)

But I run into the same issue - FormatConvertStream() won't work with BluetoothA2DPSink.   Do you have or can you make an example showing converting a2dp?  Here is my code:

```C++
#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

BluetoothA2DPSink a2dp_in;
I2SStream i2s_out;

AudioInfo from(44100, 2, 16);
AudioInfo   to(44100, 1, 16);

// Convert on Output side
FormatConverterStream converter(i2s_out);
StreamCopy copier(converter, a2dp_in);

void setup() {

    // I2S DAC settings
    auto cfg        = i2s_out.defaultConfig(TX_MODE);
   
    // ESP32 pins library default bck=14, ws=15, data=22
    cfg.pin_bck     = 15;
    cfg.pin_ws      = 27;
    cfg.pin_data    = 26;

    // mix stereo channels together into mono (both channels will get the same signal)
    a2dp_in.set_mono_downmix(true);

    // Tell I2S to send audio to only one speaker
    cfg.channels    = 1;  // (Left? can it be set to either one?)

    converter.begin(from, to);
    cfg.copyFrom(to);

    i2s_out.begin(cfg);
    a2dp_in.start("A2DP_Stereo_to_I2S_Mono");
}

void loop() {
  copier.copy();
}
```

**Reply by pschatzmann** [reply]
You can't use a copier: remove it again!
Just specify the FormatConverterStream as output with BluetoothA2DPSink a2dp_in(converter);
a2dp_in.set_mono_downmix(true); is then also not necessary because the FormatConverterStream does the mixing.

ps: code needs to be enclosed by 3 `

**Reply by fscii** [reply]
Ok thanks for all of the tips.  There is still a single issue.  Code compiles fine and it may not even be a code issue.  I still get 2 speaker output from PCM5102 (yellow and purple).  Is that a hardware issue of the PCM5102?

Also, when setting single channel, does it default to left? Can that be switched to right?  I'm curious.

```C++
#include "AudioTools.h"
#include "BluetoothA2DPSink.h"

AudioInfo from(44100, 2, 16);
AudioInfo     to(44100, 1, 16);

I2SStream i2s_out;
FormatConverterStream converter(i2s_out);
BluetoothA2DPSink a2dp_in(converter);

void setup() {

    // I2S DAC settings
    auto cfg        = i2s_out.defaultConfig(TX_MODE);
   
    // ESP32 pins library default bck=14, ws=15, data=22
    cfg.pin_bck     = 15;
    cfg.pin_ws      = 27;
    cfg.pin_data    = 26;

    // Tell I2S to send audio to only one speaker
    cfg.channels    = 1;  // (Left? can it be set to either one?)

    converter.begin(from, to);
    cfg.copyFrom(to);

    i2s_out.begin(cfg);
    a2dp_in.start("A2DP_Stereo_to_I2S_Mono");
}

void loop() {
}

**Reply by pschatzmann** [reply]
What is your ESP32 core version ?

**Reply by fscii** [reply]
2.0.16 but there is an update available to 3.0.1.  I was under the impression to hold off of 3 for now?

**Reply by pschatzmann** [reply]
In 3.0.1 it might work automatically. In 2.0.16 you can try to set the channel_format

**Reply by fscii** [reply]
I added:

```C++
cfg.channel_format = I2S_CHANNEL_FMT_ONLY_LEFT;
```
Still get audio from both speakers.  Is this not possible in Core 2?  Do I need to move to core 3?  I'm afraid it'll break other libraries I use :/

**Reply by pschatzmann** [reply]
You can give it a try: I did not consider this as an issue so far because I2S is a stereo protocol

**Reply by fscii** [reply]
SUCCESS!! by updating to Core 3 and removing: cfg.channel_format = I2S_CHANNEL_FMT_ONLY_LEFT;

However this brings more questions...  in core 2 you went through the trouble of adding: I2S_CHANNEL_FMT_ONLY_LEFT and other formats (right only etc) but they are not behaving properly?  I don't ask you to do more work to a core going away, but these options are not present for core 3?  I really like them for a project I'm working on.  Happy to donate a little bit here and there and hoping this functionality (if possible) is added in at least Core 3, but perhaps adjusted somehow for Core 2?

Any opinion on doing these things?  I'm interested and very much appreciate your work on this project and attention to questions is top notch on Github or anywhere.

Thank you very much!

**Reply by pschatzmann** [reply]
I think In the core 2 this might have been done for the ADC and not for the DAC.
In the core 3 the API is very different and I don't remember having seen any option to specify this.

Fee free to have a [look at the implementation](https://github.com/pschatzmann/arduino-audio-tools/tree/main/src/AudioI2S)

**Reply by pschatzmann** [reply]
I committed some DRAFT implementation for the 3.0.1 core: Can you double check ?
For a mono signal you can select
cfg.channel_format = I2SChannelSelect::Right; 
cfg.channel_format = I2SChannelSelect::Left;

**Reply by fscii** [reply]
Confirmed both work as expected!  Very nice!!!  Thank you!
